src/linker/utilities/spark.smk

```diff
@@ -0,0 +1,199 @@
+    &> {output} &
+    spid=$!
+
+    trap "kill -9 $spid &>/dev/null" EXIT
+
+    while true; do
+
+        if ! kill -0 $spid >/dev/null 2>&1; then
+            echo "rule $spid died"
+            cat {output} >&2
+            exit 1
+        fi
+
+        if [[ -e "{params.terminate_file_name}" ]] ; then
+            break
+        fi
+
+        sleep 1
+    done
+    """
+
+
+rule wait_for_spark_master:
+    output:
+        temp(config["results_dir"] + "/spark_logs/spark_master_uri.txt"),
+    params:
+        spark_master_log_file=rules.start_spark_master.output,
+        max_attempts=20,
+        attempt_sleep_time=10,
+    localrule: True
+    shell:
+        """
+        echo "Searching for Spark master URL in {params.spark_master_log_file}"
+        sleep {params.attempt_sleep_time}
+        num_attempts=1
+        while true; do
+
+            if [[ -e {params.spark_master_log_file} ]]; then
+                found=`grep -o "\(spark://.*$\)" {params.spark_master_log_file} || true`
+
+                if [[ ! -z $found ]]; then
+                    echo "Spark master URL found: $found"
+                    echo $found > {output}
+                    break
+                fi
+            fi
+
+            if (( $num_attempts > {params.max_attempts})); then
+                echo "Couldn't find Spark master after {params.max_attempts} attempts. Exiting."
+                exit 2
+            
+            fi
+
+            echo "Unable to find Spark master URL in logfile. Waiting {params.attempt_sleep_time} seconds and retrying..."
+            echo "(attempt $num_attempts/{params.max_attempts})"
+            sleep {params.attempt_sleep_time}
+            num_attempts=$((num_attempts + 1))
+
+        done
+        """
+
+
+rule split_workers:
+    localrule: True
+    input:
+        rules.wait_for_spark_master.output,
+    output:
+        temp(
+            touch(
+                scatter.num_workers(
+                    config["results_dir"] + "/spark_logs/spark_worker_{scatteritem}.txt"
+                )
+            )
+        ),
+
+
+rule start_spark_worker:
+    input:
+        masterurl=rules.wait_for_spark_master.output,
+        workers=config["results_dir"] + "/spark_logs/spark_worker_{scatteritem}.txt",
+    output:
+        config["results_dir"] + "/spark_logs/spark_worker_log_{scatteritem}.txt",
+    params:
+        terminate_file_name=rules.terminate_spark.output,
+        user=os.environ["USER"],
+    container:
+        "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/spark_cluster.sif"
+    shell:
+        """
+        echo " Starting Spark Worker {wildcards.scatteritem}"
+        SPARK_ROOT=/opt/spark
+        SPARK_WORKER_WEBUI_PORT=28510
+        read -r MASTER_URL < {input.masterurl}
+
+         rm -f {output} || true
+         mkdir -p "/tmp/spark_cluster_{params.user}"
+
+        $SPARK_ROOT/bin/spark-class org.apache.spark.deploy.worker.Worker \
+        --cores 1 --memory 1G \
```

Comment:
parameterize this

Reply:
You likely don't want to set the memory for the Spark worker to equal the requested memory on Slurm -- that will lead to Slurm killing it pretty often. Some buffer is needed. I can't remember how/if we dealt with this in the previous `linker` implementation but in the case study I currently leave 500MB of buffer.

Reply:
Also, I think we already have a backlog ticket for this but it would be cool if this script automatically restarted the worker process if it died (e.g. was killed by Slurm)!